fabtests/multinode/src/timing.c

```diff
@@ -0,0 +1,192 @@
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#include <string.h>
+#include <limits.h>
+#include <shared.h>
+#include "timing.h"
+#include "core.h"
+
+void multi_timer_init(struct multi_timer *timer, int rank)
+{
+	timer->rank = rank;
+	timer->start = 0;
+	timer->end = 0;
+}
+
+void multi_timer_start(struct multi_timer *timer)
+{
+	if (timer && timer->start == 0)
+		timer->start = ft_gettime_ns();
+}
+
+void multi_timer_stop(struct multi_timer *timer)
+{
+	if (timer && timer->start != 0 && timer->end == 0)
+		timer->end = ft_gettime_ns();
+}
+
+static inline void print_timer(struct multi_timer timer, char* info)
+{
+	printf("rank: %i, start: %ld, end: %ld, %s\n", 
+			timer.rank, timer.start, timer.end, info);
+}
+
+int multi_timer_analyze(struct multi_timer *timers, int timer_count)
+{
+	int i, j, ret = 0;
+	int iterations = timer_count / pm_job.num_ranks;
+	double total_timers, total_min, total_max, total_sum_time, total_duration;
+	long *min = calloc(iterations, sizeof(*min));
+	long *max = calloc(iterations, sizeof(*max));
+	long *first_start = calloc(iterations, sizeof(*first_start));
+	long *last_end = calloc(iterations, sizeof(*last_end));
+	double *sum_time = calloc(iterations, sizeof(*sum_time));
+	long duration;
+	struct multi_timer *gather_timers;
+	double iter_timer_count = 0;
+
+	gather_timers = calloc(pm_job.num_ranks * pm_job.num_ranks, 
+							sizeof(*gather_timers));
+	
+	if (!(min && max && first_start && last_end && sum_time && gather_timers))
+		goto err;
+
+	printf("%-10s %16s %16s %16s %16s\n",
+			"Iteration", "Min Send (ns)", "Max Send (ns)", 
+			"Pattern Time(ns)", "Average Send(ns)");
+
+	for (i = 0; i < iterations; i++) {
+		ret = multi_timer_iter_gather(gather_timers, timers, i);
+		if (ret < 0)
+			goto err;
+		if (pm_job.my_rank == 0) {
+			for (j = 0; j < pm_job.num_ranks * pm_job.num_ranks; j++) {
+				if (gather_timers[j].start == 0 || gather_timers[j].end == 0)
```

Comment:
Here and above, please check line lengths and limit to about 80 characters.  A couple characters over is fine if it helps with readability.